sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala

```diff
@@ -191,6 +216,216 @@ class RapidsHostMemoryStore(
+     * a `RapidsHostColumnEventHandler` object and associates it with the column's
+     * `eventHandler` or calls into the existing one, and registers itself.
+     *
+     * The registration has two goals: it accounts for repetition of a column
+     * in a `RapidsHostColumnarBatch`. If a batch has the same column repeated it must adjust
+     * the refCount at which this column is considered spillable.
+     *
+     * The second goal is to account for aliasing. If two host batches alias this column
+     * we are going to mark it as non spillable.
+     *
+     * @param rapidsHostCb - the host batch that is registering itself with this tracker
+     */
+    def register(rapidsHostCb: RapidsHostColumnarBatch, repetition: Int): Unit = {
+      registration.put(rapidsHostCb, repetition)
+    }
+
+    /**
+     * This is invoked during `RapidsHostColumnarBatch.free` in order to remove the entry
+     * in `registration`.
+     *
+     * @param rapidsHostCb - the batch that is de-registering itself
+     */
+    def deregister(rapidsHostCb: RapidsHostColumnarBatch): Unit = {
+      registration.remove(rapidsHostCb)
+    }
+
+    // called with the cudf HostColumnVector lock held from cuDF's side
+    override def onClosed(cudfCv: HostColumnVector, refCount: Int): Unit = {
+      // we only handle spillability if there is a single batch registered
+      // (no aliasing)
+      if (registration.size == 1) {
+        val (rapidsHostCb, spillableRefCount) = registration.head
+        if (spillableRefCount == refCount) {
+          rapidsHostCb.onColumnSpillable(cudfCv)
+        }
+      }
+    }
+  }
+
+  /**
+   * A `RapidsHostColumnarBatch` is the spill store holder of ColumnarBatch backed by
+   * HostColumnVector.
+   *
+   * This class owns the host batch and will close it when `close` is called.
+   *
+   * @param id the `RapidsBufferId` this batch is associated with
+   * @param batch the host ColumnarBatch we are managing
+   * @param spillPriority a starting spill priority
+   */
+  class RapidsHostColumnarBatch(
+      id: RapidsBufferId,
+      hostCb: ColumnarBatch,
+      spillPriority: Long)
+      extends RapidsBufferBase(
+        id,
+        null,
+        spillPriority) {
+
+    override val storageTier: StorageTier = StorageTier.HOST
+
+    override val needsSerialization: Boolean = true
+
+    // This is the current size in batch form. It is to be used while this
+    // batch hasn't migrated to another store.
+    private val hostSizeInByes: Long = RapidsHostColumnVector.getTotalHostMemoryUsed(hostCb)
+
+    // By default all columns are NOT spillable since we are not the only owners of
+    // the columns (the caller is holding onto a ColumnarBatch that will be closed
+    // after instantiation, triggering onClosed callbacks)
+    // This hash set contains the columns that are currently spillable.
+    private val columnSpillability = new ConcurrentHashMap[HostColumnVector, Boolean]()
```

Comment:
Nit: This could be done more efficiently as a BitSet (especially since most column batches will be less than 64 columns), although that would require synchronization when accessing it. Simply need to track the column index in the on close event handler (we're generating a new handler per column anyway) and have onColumnSpillable take a column index instead.

Reply:
Since this one is a nit, would you mind if I do this as a separate issue? This code is really a copy from the device counterpart. I would like to fix both if we are going to change this to a bitset.

Reply:
Separate issue is fine.